src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -149,11 +149,118 @@ ucs_status_t uct_rdmacm_cm_ep_conn_notify(uct_ep_h ep)
     return cep->status;
 }
 
+/**
+ * Allocate a reserved QPN either from the last FW object allocated,
+ * or by allocating a new one. When find a free QPN in an object, it
+ * always starts from last allocation position, to make sure the QPN
+ * always move forward to prevent stale QPN.
+ */
+ucs_status_t uct_rdamcm_cm_ep_create_reserved_qpn(uct_rdmacm_cm_ep_t *cep)
+{
+    struct uct_rdmacm_cm_reserved_qpn_context *ctx;
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    uint32_t qpns_per_obj;
+    ucs_status_t status;
+
+    status = uct_rdmacm_cm_get_reserved_qpn_context(uct_rdmacm_cm_ep_get_cm(cep),
+                                                    cep->id->verbs, &ctx);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    if (!ctx->support) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    qpns_per_obj = 1 << ctx->log_reserved_qpn_granularity;
+
+    pthread_mutex_lock(&ctx->mutex);
+
+    blk = ucs_list_tail(&ctx->blk_list,
+            struct uct_rdmacm_cm_reserved_qpn_blk, entry);
+    if (ucs_list_is_empty(&ctx->blk_list) ||
+        (blk->next_avail_qpn_offset >= qpns_per_obj)) {
+        status = uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_ep_get_cm(cep),
+                                                      cep->id->verbs, ctx, &blk);
+        if (status != UCS_OK) {
+            goto end;
+        }
+        ucs_list_add_tail(&ctx->blk_list, &blk->entry);
+    }
+
+    cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
+    ucs_debug("created reserved qp_num: 0x%x on rdmacm_id %p",
+              cep->qpn, cep->id);
+    UCS_BITMAP_SET(blk->bmp, blk->next_avail_qpn_offset);
+    blk->next_avail_qpn_offset++;
+
+end:
+    pthread_mutex_unlock(&ctx->mutex);
+    return status;
+}
+
+/**
+ * Deallocate a reserved QPN. The FW object is destroyed only when all QPNs
+ * in this object were used and freed.
+ */
+static void uct_rdamcm_cm_ep_destroy_reserved_qpn(uct_rdmacm_cm_ep_t *cep)
+{
+    struct uct_rdmacm_cm_reserved_qpn_context *ctx;
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk, *tmp;
+    uint32_t qpns_per_obj;
+    bool found = false;
+    ucs_status_t status;
+
+    if (cep->qpn == 0) {
+        return;
+    }
+
+    ucs_debug("cm ep destroy reserved qpn on rdmacm_id %p", cep->id);
+
+    status = uct_rdmacm_cm_get_reserved_qpn_context(uct_rdmacm_cm_ep_get_cm(cep),
+                                                    cep->id->verbs, &ctx);
+    if (status != UCS_OK) {
+        ucs_error("get reserved qpn context failed");
+        return;
+    }
+
+    qpns_per_obj = 1 << ctx->log_reserved_qpn_granularity;
+
+    pthread_mutex_lock(&ctx->mutex);
+
+    ucs_list_for_each_safe(blk, tmp, &ctx->blk_list, entry) {
+        if ((cep->qpn >= blk->first_qpn) &&
+            (cep->qpn < blk->first_qpn + qpns_per_obj)) {
+            found = true;
+            break;
+        }
+    }
```

Comment:
- why `*_safe`? as far as I can see this is under lock and the list in not modified inside the loop
 - maybe move to separate func?

Reply:
@evgeny-leksikov 

- Yes, you are right. No need _safe here. Fixed.
- Prefer not move to seperate func since it's actually only 3 lines of code and it's just called only once. No need move to separate func to increase code.